Carbon nrf51

mbolivar opened the pull request:
- Add a SPIS (SPI slave) driver for nRF51 and nRF52
- Add board support for the nRF51 chip on the 96Boards Carbon

Future work will add a sample application which uses this board to provide Zephyr-based Bluetooth connectivity to the main STM32 chip on Carbon via SPI. This is currently in https://github.com/Linaro/zephyr/commit/e2d56d4a5cb0aa6e2aff6e4d5082719ff8416872 and being prepared for submission upstream.

rsalveti committed:
boards: 96b_carbon_nrf51: add support for 96Boards Carbon nRF51 chip

This adds support for the nRF51 chip on the board.

If you'd like to run Zephyr on the STM32F4 chip on Carbon, you need to
use the 96b_carbon board instead.

The current SPI Bluetooth protocol only uses 5 wires, so we use the
remaining pin as UART TX.

Signed-off-by: Ricardo Salveti <ricardo.salveti@linaro.org>
Signed-off-by: Amit Kucheria <amit.kucheria@linaro.org>
Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
drivers: spi: add nRF5 slave driver

SPIS driver for nRF51 and nRF52 SoCs.

Signed-off-by: Amit Kucheria <amit.kucheria@linaro.org>
Signed-off-by: Ricardo Salveti <ricardo.salveti@linaro.org>
Signed-off-by: Michael Scott <michael.scott@linaro.org>
Signed-off-by: Tyler Baker <tyler.baker@linaro.org>
Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

dbkinder reviewed (dismissed):
+1 for docs (thanks for not misspelling "its" anywhere :)

pfalcon reviewed (approved):
Eyeballed.

cvinayak commented on boards/arm/96b_carbon_nrf51/doc/96b_carbon_nrf51.rst:
```diff
@@ -0,0 +1,113 @@
+********
+
+Zephyr applications use the 96b_carbon_nrf51 configuration to run on
+the secondary nRF51832 chip on the 96Boards Carbon. This chip provides
```
nRF51822 / nRF51422 / nRF51824 ?  there is no nRF51832

mbolivar replied:
Thanks! That's a typo; 51822 is correct. (This is right in the DTS; I just fat-fingered the docs).

cvinayak commented on boards/arm/96b_carbon_nrf51/doc/96b_carbon_nrf51.rst:
```diff
@@ -0,0 +1,113 @@
+As an aid to debugging, this board configuration directs a console
+output to a currently unused pin. Users who are experienced in
+electronics rework can remove a resistor (R22) on the board and attach
+a wire to the nRF51832's UART output.
```
Same here ... nRF51822, right?

mbolivar commented:
Updated.

- rebase onto master
- fix docs typos

cvinayak commented on drivers/spi/spis_nrf5.c:
```diff
@@ -0,0 +1,467 @@
+typedef void (*spis_nrf5_config_t)(void);
+
+struct spis_nrf5_config {
+	NRF_SPIS_Type *regs;			/* Registers */
```
use spaces in between instead of tabs to get comment alignment consistent all over this file.

cvinayak commented on drivers/spi/spis_nrf5.c:
```diff
@@ -0,0 +1,467 @@
+{
+	__unused NRF_SPIS_Type *regs = SPI_REGS(dev);
+	__unused u32_t sck, miso, mosi, csn,
+				rxd_ptr, rxd_max, rxd_amount,
```
is the alignment correct?

cvinayak commented on drivers/spi/spis_nrf5.c:
```diff
@@ -0,0 +1,467 @@
+	SYS_LOG_DBG("SPI Slave driver init: %p", dev);
+
+	/* Enable constant latency for faster SPIS response */
+	NRF_POWER->TASKS_CONSTLAT = 1;
```
This should be a new driver (power?), bind and enable; ideally Kconfig selectable.

carlescufi replied:
I suggest we leave it as-is for this driver for now. This has been in the works for long enough.

cvinayak reviewed (changes requested):
Did not see this file, as GH hides large diffs.

mbolivar commented:
Thanks for the additional reviews.

- Rebased onto master
- Fixed style issues (@cvinakak, one overlong line)
- Used SPDX one place where it was missing

Regarding NRF_POWER, I can either leave it as is or just delete that line for now and we can continue to carry that part out of tree for the time being.

nashif commented on boards/arm/96b_carbon_nrf51/doc/96b_carbon_nrf51.rst:
```diff
@@ -0,0 +1,113 @@
+the secondary nRF51822 chip on the 96Boards Carbon. This chip provides
+Bluetooth functionality to the main STM32F401RET chip via SPI.
+
+See :ref:`96b_carbon_board` for a general overview of the board.
```
can you please follow the template in doc/templates/board.tmpl to keep board documentation consistent? Otherwise looks good

mbolivar commented:
Thanks for the pointer; I didn't know about that template and was imitating 96b_carbon and 96b_nitrogen. Now fixed.

nashif commented on boards/arm/96b_carbon_nrf51/doc/96b_carbon_nrf51.rst:
```diff
@@ -0,0 +1,136 @@
+   https://github.com/blacksphere/blackmagic/wiki/Getting-Started
+
+.. _Board documentation from 96Boards`_:
+   http://www.96boards.org/product/carbon/
```
this seems to be wrong

mbolivar replied:
What is the issue? I can load the URL from here.

nashif replied:
look here: https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/524/consolidateTests

dbkinder replied:
Problem is the extra underscore at the end and before the colon.
Should be:
     .. _Board documentation from 96Boards`:

dbkinder replied:
oops, and without the back quote either, sigh...
   .. _Board documentation from 96Boards:

mbolivar replied:
D'oh. Thanks; fixed.

nashif commented:
> Thanks for the pointer; I didn't know about that template and was imitating 96b_carbon and 96b_nitrogen. Now fixed.

Yeah, we need to fix those when we have a chance
